File: nicegui/elements/scene/objects/curve.js
import SceneLib from "nicegui-scene";
const { THREE } = SceneLib;

export default class Curve {
    create_mesh(start, control1, control2, end, num_points) {
        const curve = new THREE.CubicBezierCurve3(
            new THREE.Vector3(...start),
            new THREE.Vector3(...control1),
            new THREE.Vector3(...control2),
            new THREE.Vector3(...end),
        );
        const geometry = new THREE.BufferGeometry().setFromPoints(curve.getPoints(num_points - 1));
        const material = new THREE.LineBasicMaterial({ transparent: true });
        return new THREE.Line(geometry, material);
    }
}

File: nicegui/elements/scene/objects/curve.py
from ..scene_object3d import Object3D


class Curve(Object3D, component='curve.js'):
    def __init__(
        self,
        start: list[float],
        control1: list[float],
        control2: list[float],
        end: list[float],
        num_points: int = 20,
    ) -> None:
        """Curve

        This element is based on Three.js' `CubicBezierCurve3 <https://threejs.org/docs/index.html#api/en/extras/curves/CubicBezierCurve3>`_ object.

        :param start: start point of the curve
        :param control1: first control point of the curve
        :param control2: second control point of the curve
        :param end: end point of the curve
        :param num_points: number of points to use for the curve (default: 20)
        """
        super().__init__(start, control1, control2, end, num_points)
